feat(security): Create doc about CORS settings (#623)
edgexfoundry/edgex-docs: docs_src/security/Ch-CORS-Settings.md

# CORS enabling

!!! edgey "EdgeX 2.1"
    New for EdgeX 2.1 is the ability to enable CORS access to EdgeX microservices through configuration. 

> The EdgeX microservices provide REST APIs and those services might be called from a GUI through a browser. Most browsers prevent service calls from a different origin. Thus, EdgeX supports Cross-Origin Resource Sharing (CORS) since Jakarta release (v2.1), and this feature can be controlled by the configurations. The default behavior of CORS is disabled. Here is a good reference to understand [CORS](https://www.html5rocks.com/en/tutorials/cors//).

Reviewer: ```suggestion
The EdgeX microservices provide REST APIs and those services might be called from a GUI through a browser.  Browsers prevent service calls from a different origin, making it impossible to host a management GUI on one domain that manages an EdgeX device on a different domain. Thus, EdgeX supports Cross-Origin Resource Sharing (CORS) since Jakarta release (v2.1), and this feature can be controlled by the configurations. The default behavior of CORS is disabled. Here is a good reference to understand [CORS](https://www.html5rocks.com/en/tutorials/cors/).
```

Author: @bnevis-i the page of https://www.html5rocks.com/en/tutorials/cors// and https://www.html5rocks.com/en/tutorials/cors/ are different

Reviewer: @cloudxxx8 The final "/" on the URL is redundant.  I tried both, and both are redirecting to https://web.dev/cross-origin-resource-sharing/ for me.

!!! note
    C Device SDK doesn't support CORS, and enabling CORS in Device Services is not recommended because browsers should not access Device Services directly.

> 

Reviewer: ```suggestion

## Enabling CORS

There are two different ways to enable CORS depending on whether EdgeX is deployed in the security-enabled configuration.
In the non-security configuration, EdgeX microservices are directly exposed on host ports.
EdgeX microservices receive client requests directly in this configuration, and thus,
the EdgeX microservices themselves must respond to CORS requests.
In the security-enabled configuration,
EdgeX microservices are exposed behind an API gateway
that will receive CORS requests first.
Only authenticated calls will be forwarded to the EdgeX microservice,
but CORS pre-flight requests are always unauthenticated.

CORS can be enabled at the API gateway in a security-enabled configuration,
and at the individual microservice level in the non-security configuration.
However, implementers should choose one or the other, not both.

### Enabling CORS for API Gateway

Configure CORS in the `CORSConfiguration` configuration section for the `security-proxy-setup` microservice.
They can also be set via `CORSConfiguration_*` environment variables.
See the table below for documentation of specific parameters.

### Enabling CORS for Individual Microservices

Configure CORS in the `Service.CORSConfiguration` configuration section for each microservice to be exposed via CORS.  They can also be set via `Service_CORSConfiguration_*` environment variables.
See the table below for documentation of specific parameters.

Note that the name of the configuration section and environment variable override is intentionally different than the API gateway section,
in alignment with the guidance that CORS should be enabled at the microservice level
or the API gateway level, but not both.

```

## CORS configurations

The CORS settings are under `Service.CORSConfiguration` configuration, and they can be set via `Service_CORSConfiguration_*` environment variables.

> === "Service.CORSConfiguration"

Reviewer: ```suggestion
```

|Property|Default Value|Description|
|---|---|---|
|||The settings of controling CORS http headers|
|EnableCORS|false|Enable or disable CORS support.|
|CORSAllowCredentials|false|The value of `Access-Control-Allow-Credentials` http header. It appears only if the value is `true`.|
|CORSAllowedOrigin|"https://localhost"|The value of `Access-Control-Allow-Origin` http header.|
|CORSAllowedMethods|"GET, POST, PUT, PATCH, DELETE"|The value of `Access-Control-Allow-Methods` http header.|
|CORSAllowedHeaders|"Authorization, Accept, Accept-Language, Content-Language, Content-Type, X-Correlation-ID"|The value of `Access-Control-Allow-Headers` http header.|
|CORSExposeHeaders|"Cache-Control, Content-Language, Content-Length, Content-Type, Expires, Last-Modified, Pragma, X-Correlation-ID"|The value of `Access-Control-Expose-Headers` http header.|

> |CORSMaxAge|3600|The value of `Access-Control-Max-Age` http header.|

Reviewer 2: This section should be add to the common configuration doc and referenced from here.

Author: moved this section to common configuration page

> To understand more details about those http headers, please refer to [MDN Web Docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#the_http_response_headers).

Reviewer: ```suggestion
To understand more details about these HTTP headers, please refer to [MDN Web Docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#the_http_response_headers).
```

the EdgeX microservices themselves must respond to CORS requests.
In the security-enabled configuration,
EdgeX microservices are exposed behind an API gateway
that will receive CORS requests first.
Only authenticated calls will be forwarded to the EdgeX microservice,
but CORS pre-flight requests are always unauthenticated.

CORS can be enabled at the API gateway in a security-enabled configuration,
and at the individual microservice level in the non-security configuration.
However, implementers should choose one or the other, not both.

### Enabling CORS for Individual Microservices

Configure CORS in the `Service.CORSConfiguration` configuration section for each microservice to be exposed via CORS.  They can also be set via `Service_CORSConfiguration_*` environment variables.
Please refer to the [Common Configuration](../microservices/configuration/CommonConfiguration.md/#configuration-properties) page to learn the details.

### Enabling CORS for API Gateway

Configure CORS in the `CORSConfiguration` configuration section for the `security-proxy-setup` microservice.
They can also be set via `CORSConfiguration_*` environment variables.

> Please refer to the [Common Configuration](../microservices/configuration/CommonConfiguration.md/#configuration-properties) page to learn the details.

Reviewer 2: Please make this `!!! note` with additional text stating `The settings under the `CORSConfiguration` configuration section are the same as those under the Service.CORSConfiguration so please refer ...`

> Note that the name of the configuration sections and environment variable overrides are intentionally different than the API gateway section, in alignment with the guidance that CORS should be enabled at the microservice level or the API gateway level, but not both.

Reviewer 2: Please make this a `!!! note`

## Enabling CORS

There are two different ways to enable CORS depending on whether EdgeX is deployed in the security-enabled configuration.
In the non-security configuration, EdgeX microservices are directly exposed on host ports.
EdgeX microservices receive client requests directly in this configuration, and thus,
the EdgeX microservices themselves must respond to CORS requests.
In the security-enabled configuration,
EdgeX microservices are exposed behind an API gateway
that will receive CORS requests first.
Only authenticated calls will be forwarded to the EdgeX microservice,
but CORS pre-flight requests are always unauthenticated.

CORS can be enabled at the API gateway in a security-enabled configuration,
and at the individual microservice level in the non-security configuration.
However, implementers should choose one or the other, not both.

### Enabling CORS for API Gateway

Configure CORS in the `CORSConfiguration` configuration section for the `security-proxy-setup` microservice.
They can also be set via `CORSConfiguration_*` environment variables.

> See the table below for documentation of specific parameters.

Reviewer 2: No table below. Change to reference CORS configurations section below

### Enabling CORS for Individual Microservices

Configure CORS in the `Service.CORSConfiguration` configuration section for each microservice to be exposed via CORS.  They can also be set via `Service_CORSConfiguration_*` environment variables.

> See the table below for documentation of specific parameters.

Reviewer 2: No table below. Change to reference CORS configurations section below

> Note that the name of the configuration section and environment variable override is intentionally different than the API gateway section,

Reviewer 2: ```suggestion
Note that the name of the configuration sections and environment variable overrides are intentionally different than the API gateway section,
```

> Note that the name of the configuration section and environment variable override is intentionally different than the API gateway section,

Reviewer 2: IMO, this note should be at the bottom after reference to common config.

in alignment with the guidance that CORS should be enabled at the microservice level
or the API gateway level, but not both.

### CORS configurations

The CORS settings are under `Service.CORSConfiguration` configuration, and they can be set via `Service_CORSConfiguration_*` environment variables.

> Please refer to the [Common Configuration](../microservices/configuration/CommonConfiguration.md) page to learn the details.

Reviewer 2: Add note above that individual settings under `Service.CORSConfiguration` and `CORSConfiguration` are the same, so use common Service.CORSConfiguration as reference.

Reviewer 2: ```suggestion
Please refer to the [Common Configuration](../microservices/configuration/CommonConfiguration.md/#configuration-properties) page to learn the details.
```
